plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,220 @@
+
+func (p *Parser) Init() error {
+	if (p.Schema == "" && p.SchemaRegistry == "") || (p.Schema != "" && p.SchemaRegistry != "") {
+		return errors.New("exactly one of 'schema_registry' or 'schema' must be specified")
+	}
+	if p.TimestampFormat == "" {
+		return errors.New("must specify 'timestamp_format'")
+	}
+	if p.SchemaRegistry != "" {
+		p.registryObj = newSchemaRegistry(p.SchemaRegistry)
+	}
+	return nil
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+	var codec *goavro.Codec
+	var err error
+	var message []byte
+	message = buf[:]
+
+	if p.registryObj != nil {
+		// The input must be Confluent Wire Protocol
+		if buf[0] != 0 {
+			return nil, errors.New("first byte is not 0: not Confluent Wire Protocol")
+		}
+		schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+		schemastruct, err := p.registryObj.getSchemaAndCodec(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = schemastruct.Schema
+		codec = schemastruct.Codec
+		message = buf[5:]
+	} else {
+		// Check for single-object encoding
+		magicBytes := int(binary.BigEndian.Uint16(buf[:2]))
+		expectedMagic := int(binary.BigEndian.Uint16([]byte("c301")))
+		if magicBytes == expectedMagic {
+			message = buf[10:]
+			// We could in theory validate the fingerprint against
+			// the schema.  Maybe later.
+			// We would get the fingerprint as int(binary.LittleEndian.Uint64(buf[2:10]))
+		} // Otherwise we assume bare Avro binary
+		schema = p.Schema
+		codec, err = goavro.NewCodec(schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	native, _, err := codec.NativeFromBinary(message)
+	if err != nil {
+		return nil, err
+	}
+	// Cast to string-to-interface
+	codecSchema, ok := native.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("native is of unsupported type %T", native)
+	}
+	m, err := p.createMetric(codecSchema, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, errors.New("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) createMetric(data map[string]interface{}, schema string) (telegraf.Metric, error) {
+	// Tags differ from fields, in that tags are inherently strings.
+	// fields can be of any type.
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	// Set default tag values
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+	// Avro doesn't have a Tag/Field distinction, so we have to tell
+	// Telegraf which items are our tags.
+	for _, tag := range p.Tags {
+		sTag, err := internal.ToString(data[tag])
+		if err != nil {
+			p.Log.Warnf("Could not convert %v to string for tag %q: %v", data[tag], tag, err)
+		}
+		tags[tag] = sTag
```

Comment:
If we failed to convert, we shouldn't try to assign either (i.e. if err: warn + continue)